source/server/configuration_impl.cc

```diff
@@ -128,7 +128,10 @@ absl::Status MainImpl::initialize(const envoy::config::bootstrap::v3::Bootstrap&
   }
 
   ENVOY_LOG(info, "loading {} cluster(s)", bootstrap.static_resources().clusters().size());
+
+  // clusterManagerFromProto() and init() have to be called consecutively.
   cluster_manager_ = cluster_manager_factory.clusterManagerFromProto(bootstrap);
+  RETURN_IF_NOT_OK(cluster_manager_->init(bootstrap));
```

Comment:
Not sure if this completely resolve our problem. This ensure the server context could always return a valid reference to the cluster manager. But what will happen when the upstream network filter or upstream http filter try to access the actual content of the cluster manager when doing the initialization?

Reply:
Thanks for the comments!

1) When router filter is trying to call processFilters() for upstream filters:
https://github.com/envoyproxy/envoy/blob/63bc9b564b1a76a22a0d029bcac35abeffff2a61/source/common/router/router.cc#L108
it does access context.serverFactoryContext().clusterManager().  IIUC, this is in the much later stage, so the cluster manager instance is already initialized.

2) In the ClusterInfoImpl c'tor, which is called during cluster_manager_->init()
https://github.com/envoyproxy/envoy/blob/63bc9b564b1a76a22a0d029bcac35abeffff2a61/source/common/upstream/upstream_impl.cc#L1376

https://github.com/envoyproxy/envoy/blob/63bc9b564b1a76a22a0d029bcac35abeffff2a61/source/common/upstream/cluster_manager_impl.cc#L905

It's accessing the cluster_manager_, before it is initialized.  This means ClusterInfoImpl c'tor only needs an instance of cluster manager,  and does not require it is initialized.

Thus the change in this PR should be safe.  BTW, there is no actual logical change in this PR.